standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStoreClientWithLocalCache.java

```diff
@@ -0,0 +1,246 @@
+   * @return Result object / null
+   * @throws TException
+   */
+  private Object getResultObject(CacheKey cacheKey) throws TException {
+    Object result = null;
+
+    switch (cacheKey.IDENTIFIER) {
+      case PARTITIONS_BY_EXPR:
+        result = super.getPartitionsByExprResult((PartitionsByExprRequest)cacheKey.obj);
+        break;
+      case PARTITIONS_SPEC_BY_EXPR:
+        result = super.getPartitionsSpecByExprResult((PartitionsByExprRequest)cacheKey.obj);
+        break;
+      default:
+        break;
+    }
+
+    return result;
+  }
+
+  @Override
+  protected PartitionsByExprResult getPartitionsByExprResult(PartitionsByExprRequest req) throws TException {
+    PartitionsByExprResult r = null;
+
+    // table should be transactional to get responses from the cache
+    if (IS_CACHE_ENABLED && isRequestCachable(req)) {
+      CacheKey cacheKey = new CacheKey(KeyType.PARTITIONS_BY_EXPR, req);
+      Object val;
+      try {
+        val = mscLocalCache.get(cacheKey, this::load); // get either the result or an Exception
+
+        if (val instanceof PartitionsByExprResult) {
+          r = (PartitionsByExprResult) val;
+          LOG.debug(cacheObjName + ": " + mscLocalCache.stats().toString());
+        } else if (val instanceof Exception) {
+          mscLocalCache.invalidate(cacheKey);
+          throw (Exception)val;
+        }
+      } catch (MetaException me) {
+        throw me;
+      } catch (Exception e) {
+        LOG.error("Exception in MSC local cache: {}", e.toString());
+        throw new TException(e.getMessage());
+      }
+    } else {
+         r = client.get_partitions_by_expr(req);
+    }
+
+    return r;
+  }
+
+  @Override
+  protected PartitionsSpecByExprResult getPartitionsSpecByExprResult(PartitionsByExprRequest req) throws TException {
+    PartitionsSpecByExprResult r = null;
+
+    // table should be transactional to get responses from the cache
+    if (IS_CACHE_ENABLED && isRequestCachable(req)) {
+      CacheKey cacheKey = new CacheKey(KeyType.PARTITIONS_SPEC_BY_EXPR, req);
+      Object val;
+      try {
+        val = mscLocalCache.get(cacheKey, this::load);
+
+        if (val instanceof PartitionsSpecByExprResult) {
+          r = (PartitionsSpecByExprResult) val;
+          LOG.debug(cacheObjName + ": " + mscLocalCache.stats().toString());
+        } else if (val instanceof Exception) {
+          mscLocalCache.invalidate(cacheKey);
+          throw (Exception)val;
+        }
+      } catch (MetaException me) {
+        throw me;
+      } catch (Exception e) {
+        LOG.error("Exception in MSC local cache: {}", e.toString());
+        throw new TException(e.getMessage());
+      }
+    } else {
+        r = client.get_partitions_spec_by_expr(req);
+    }
+
+    return r;
+  }
+
+  /**
+   * This method determines if the request should be cached.
+   * @param request Request object
+   * @return boolean
+   */
+  private boolean isRequestCachable(Object request) {
```

Comment:
Can we pass the `KeyType` as parameter too and check based on switch statement on the enum?